pallets/proxy-keystore/src/lib.rs

```diff
@@ -0,0 +1,311 @@
+		/// Add keys to the storages.
+		#[pallet::weight(T::WeightInfo::add_keys(T::MaxKeys::get() as u32))]
+		pub fn add_keys(origin: OriginFor<T>, keys: Vec<AddKey<T::Hash>>) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+
+			ensure!(keys.len() > 0, Error::<T>::NoKeys);
+			ensure!(
+				keys.len() <= T::MaxKeys::get() as usize,
+				Error::<T>::TooManyKeys
+			);
+
+			let key_deposit = <KeyDeposit<T>>::get();
+
+			for add_key in keys {
+				Self::add_key(account_id.clone(), add_key.clone(), key_deposit.clone())?;
+			}
+
+			Ok(())
+		}
+
+		/// Remove keys from the storages.
+		#[pallet::weight(T::WeightInfo::revoke_keys(T::MaxKeys::get() as u32))]
+		pub fn revoke_keys(origin: OriginFor<T>, key_hashes: Vec<T::Hash>) -> DispatchResult {
+			let account_id = ensure_signed(origin)?;
+
+			ensure!(key_hashes.len() > 0, Error::<T>::NoKeys);
+			ensure!(
+				key_hashes.len() <= T::MaxKeys::get() as usize,
+				Error::<T>::TooManyKeys
+			);
+
+			for key_hash in key_hashes {
+				Self::revoke_key(account_id.clone(), key_hash.clone())?;
+			}
+
+			Ok(())
+		}
+
+		/// Set a new key deposit.
+		#[pallet::weight(T::WeightInfo::set_deposit())]
+		pub fn set_deposit(origin: OriginFor<T>, new_deposit: T::Balance) -> DispatchResult {
+			T::AdminOrigin::ensure_origin(origin)?;
+
+			<KeyDeposit<T>>::set(new_deposit);
+
+			Self::deposit_event(Event::DepositSet(new_deposit));
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Add a key to the `Keys` and `LastKeyByPurpose` storages if the following checks pass:
+		///
+		/// - The account has enough funds to cover the `key_deposit`;
+		/// - A key with the same hash and purpose does not exist in the `Keys` storage;
+		///
+		/// The `key_deposit` is reserved upon success.
+		fn add_key(
+			account_id: T::AccountId,
+			add_key: AddKey<T::Hash>,
+			key_deposit: T::Balance,
+		) -> DispatchResult {
+			T::Currency::reserve(&account_id, key_deposit)?;
+
+			let key_id: KeyId<T::Hash> = (add_key.key.clone(), add_key.purpose.clone());
+
+			<Keys<T>>::try_mutate(
+				account_id.clone(),
+				key_id.clone(),
+				|key_opt| -> DispatchResult {
+					match key_opt {
+						Some(_) => Err(Error::<T>::KeyAlreadyExists.into()),
+						None => {
+							let _ = key_opt.insert(Key {
+								purpose: add_key.purpose.clone(),
+								key_type: add_key.key_type.clone(),
+								revoked_at: None,
+								deposit: key_deposit,
+							});
+
+							Ok(())
+						}
+					}
+				},
+			)?;
+
+			<LastKeyByPurpose<T>>::insert(account_id.clone(), add_key.purpose.clone(), key_id);
+
+			Self::deposit_event(Event::KeyAdded(
+				account_id.clone(),
+				add_key.key.clone(),
+				add_key.purpose.clone(),
+				add_key.key_type.clone(),
+			));
+
+			Ok(())
+		}
+
+		/// Revoke a key at the current `block_number` in the `Keys` storage
+		/// if the key is found and it's *not* already revoked.
+		fn revoke_key(account_id: T::AccountId, key: T::Hash) -> DispatchResult {
```

Comment:
Here we should allow a user to pass the key purpose, that way we enable the option to revoke keys per purpose, and not revoking all of them. 
```suggestion
		fn revoke_key(account_id: T::AccountId, key_id: KeyId<T::Hash>) -> DispatchResult {
```
And internally is just a direct mutate with pretty much the same logic you already have but erroring our in the `None` branch with `KeyNotFound`

Reply:
Done, PTAL.